### ISSUE TITLE
Native secure websocket hangs on test fleet

### ISSUE THREAD

**Issue by D4nte** [status-im/nim-waku#845]
It is not possible to connect to a test fleet node via websocket (using nim-waku's native websocket feature).

Using `websocat`:

```
▶ websocat -v wss://node-01.ac-cn-hongkong-c.wakuv2.test.statusim.net   # Websockify, works fine
[INFO  websocat::lints] Auto-inserting the line mode
[INFO  websocat::stdio_threaded_peer] get_stdio_peer (threaded)
[INFO  websocat::ws_client_peer] get_ws_client_peer
[INFO  websocat::ws_client_peer] Connected to ws
/multistream/1.0.0
^C

▶ websocat -v wss://node-01.ac-cn-hongkong-c.wakuv2.test.statusim.net:8000 # Native websocket, hangs
[INFO  websocat::lints] Auto-inserting the line mode
[INFO  websocat::stdio_threaded_peer] get_stdio_peer (threaded)
[INFO  websocat::ws_client_peer] get_ws_client_peer
```

In chrome, we have the following stack (not sure it helps)

```
client.js:18 WebSocket connection to 'wss://node-01.do-ams3.wakuv2.test.statusim.net:8000/p2p/16Uiu2HAmPLe7Mzm8TsYUubgCAW1aJoeFScxrLj8ppHFivPo97bUZ' failed: WebSocket is closed before the connection is established.
(anonymous) @ client.js:18
stream.close @ client.js:16
(anonymous) @ index.js:103
setTimeout (async)
onAbort @ index.js:102
onAbort @ index.js:13
onAbort @ index.js:13
abort @ index.js:26
(anonymous) @ index.js:16
Retimer._timerWrapper @ retimer.js:21
setTimeout (async)
Retimer @ retimer.js:25
retimer @ retimer.js:77
TimeoutController @ index.js:16
_createPendingDial @ index.js:281
connectToPeer @ index.js:157
await in connectToPeer (async)
_maybeConnect @ index.js:746
(anonymous) @ index.js:691
emit @ events.js:153
add @ address-book.js:301
await in add (async)
_onDiscoveryPeer @ index.js:727
emit @ events.js:153
(anonymous) @ index.js:73
_discoverBootstrapPeers @ index.js:57
start @ index.js:46
(anonymous) @ index.js:809
_setupPeerDiscovery @ index.js:809
_onDidStart @ index.js:706
await in _onDidStart (async)
start @ index.js:389
await in start (async)
(anonymous) @ waku.ts:227
fulfilled @ utils.ts:39
Promise.then (async)
step @ utils.ts:39
fulfilled @ utils.ts:39
Promise.then (async)
step @ utils.ts:39
(anonymous) @ utils.ts:39
__awaiter @ utils.ts:39
create @ waku.ts:151
(anonymous) @ App.js:25
invokePassiveEffectCreate @ react-dom.development.js:23487
callCallback @ react-dom.development.js:3945
invokeGuardedCallbackDev @ react-dom.development.js:3994
invokeGuardedCallback @ react-dom.development.js:4056
flushPassiveEffectsImpl @ react-dom.development.js:23574
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
flushPassiveEffects @ react-dom.development.js:23447
(anonymous) @ react-dom.development.js:23324
workLoop @ scheduler.development.js:417
flushWork @ scheduler.development.js:390
performWorkUntilDeadline @ scheduler.development.js:157
```

Trying out with local nim-waku node now.

**Comment by jm-clius** [comment]
I'll take a look at this this week. Do you know if it's just related to the certificate domain: https://github.com/status-im/nim-waku/pull/848. Need to fix this ASAP, probably by just adding support for `dns4` addrs in `nwaku`.

**Comment by D4nte** [comment]
> I'll take a look at this this week. Do you know if it's just related to the certificate domain: #848. Need to fix this ASAP, probably by just adding support for `dns4` addrs in `nwaku`.

I do not think it's related to #848 as I am using the fqdn to reach the host and nim-waku is configured to used letsencrypt fqdn certs for the secure websocket connection.
It's weird because I tried on my own box with nim-waku master with letsencrypt certs and it was fine.

**Comment by jm-clius** [comment]
I am unable to reach the websocket port `8000` at all. @jakubgs this port is included in the [open_ports_list](https://github.com/status-im/infra-nim-waku/blob/master/ansible/group_vars/wakuv2-test.yml#L52-L59) but I'm not sure if this is enough for it to be reachable?

The websockify port is reachable and works:

```shell
telnet node-01.do-ams3.wakuv2.test.statusim.net 443
Trying 134.209.139.210...
Connected to node-01.do-ams3.wakuv2.test.statusim.net.
```

But the native port:

```shell
telnet node-01.do-ams3.wakuv2.test.statusim.net 8000
Trying 134.209.139.210...
telnet: Unable to connect to remote host: Connection timed out
```

**Comment by jakubgs** [comment]
Indeed, it is filtered:
```
 > sudo nmap -Pn -p8000 node-01.do-ams3.wakuv2.test  
Starting Nmap 7.92 ( https://nmap.org ) at 2022-02-22 12:18 CET
Nmap scan report for node-01.do-ams3.wakuv2.test (134.209.139.210)
Host is up.

PORT     STATE    SERVICE
8000/tcp filtered http-alt

Nmap done: 1 IP address (1 host up) scanned in 2.20 seconds
```

**Comment by jakubgs** [comment]
But it is opened on the firewall:
```
admin@node-01.do-ams3.wakuv2.test:~ % t -n -L SERVICES
Chain SERVICES (2 references)
target     prot opt source               destination         
ACCEPT     tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:22 /* SSH access */
ACCEPT     udp  --  0.0.0.0/0            0.0.0.0/0            udp dpt:51820 /* WireGuard VPN UDP */
ACCEPT     tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:30303 /* nim-waku-v2 */
ACCEPT     udp  --  0.0.0.0/0            0.0.0.0/0            udp dpt:30303 /* nim-waku-v2 */
ACCEPT     tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:443 /* nim-waku-v2 */
ACCEPT     tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8000 /* websocket test */
ACCEPT     tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:80 /* Certbot verification */
ACCEPT     tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8000 /* nim-waku-v2 */
```

**Comment by jakubgs** [comment]
I can see the port is mapped without and address and listens on `0.0.0.0`:
```
admin@node-01.do-ams3.wakuv2.test:/docker/nim-waku-v2 % grep 8000 docker-compose.yml
      - '8000:8000/tcp'
      --websocket-port=8000

admin@node-01.do-ams3.wakuv2.test:/docker/nim-waku-v2 % sudo netstat -lpnt | grep 8000
tcp        0      0 0.0.0.0:8000            0.0.0.0:*               LISTEN      761/dockerd    
```

**Comment by jakubgs** [comment]
We definitely have it open on the cloud provider firewall:
```tf
  /* firewall */
  open_tcp_ports = [
    "80",    /* certbot */
    "443",   /* p2p websockify */
    "8000",  /* p2p websocket */
```
https://github.com/status-im/infra-nim-waku/blob/9676e4e8237427552a3c52aec9bfd14163b98dbf/hosts.tf#L18-L22

**Comment by jakubgs** [comment]
Interesting, after using `docker-compose up --force-recreate -d` it started working:
```
 > sudo nmap -Pn -p8000 node-01.do-ams3.wakuv2.test
Starting Nmap 7.92 ( https://nmap.org ) at 2022-02-22 12:47 CET
Nmap scan report for node-01.do-ams3.wakuv2.test (134.209.139.210)
Host is up (0.12s latency).

PORT     STATE SERVICE
8000/tcp open  http-alt

Nmap done: 1 IP address (1 host up) scanned in 0.19 seconds
```
Seems like the same issue described in https://github.com/status-im/infra-waku-connect/issues/1#issuecomment-1047687494.

Seems like the port stops responding the TCP handshakes after some time.

**Comment by jakubgs** [comment]
For example, `node-01.ac-cn-hongkong-c.wakuv2.test` still appears as filtered:
```
Nmap scan report for node-01.ac-cn-hongkong-c.wakuv2.test (47.242.210.73)
Host is up.

PORT     STATE    SERVICE
8000/tcp filtered http-alt
```
But the node is up and healthy:
```
admin@node-01.ac-cn-hongkong-c.wakuv2.test:~ % d
CONTAINER ID   NAMES         IMAGE                                CREATED        STATUS
d46d8c291795   nim-waku-v2   statusteam/nim-waku:deploy-v2-test   6 days ago     Up 6 days (healthy)
263c194bbb0b   websockify    statusteam/websockify:v0.10.0        4 months ago   Up 36 hours
```
And if I try to access it locally it fails:
```
admin@node-01.ac-cn-hongkong-c.wakuv2.test:~ % websocat -vk1E wss://localhost:8000 
[INFO  websocat::lints] Auto-inserting the line mode                             
[INFO  websocat::stdio_threaded_peer] get_stdio_peer (threaded)
[INFO  websocat::ws_client_peer] get_ws_client_peer
websocat: WebSocketError: I/O failure
websocat: error running
```
So something is definitely up.

**Comment by jakubgs** [comment]
If I do the `websocat` test while I have `tcpdump` running I see this:
```
admin@node-01.ac-cn-hongkong-c.wakuv2.test:~ % sudo tcpdump -i lo
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on lo, link-type EN10MB (Ethernet), capture size 262144 bytes
11:53:23.219086 IP6 localhost.36892 > localhost.8000: Flags [S], seq 771356695, win 65476, options [mss 65476,sackOK,TS val 355800288 ecr 0,nop,wscale 7], length 0
11:53:23.219099 IP6 localhost.8000 > localhost.36892: Flags [R.], seq 0, ack 771356696, win 0, length 0
```
So the `8000` port is returning a `RST` packet and closes the connection.

**Comment by jakubgs** [comment]
Interestingly enough, if I try to connect from outside I don't get a `RST`, I get nothing:
```
admin@node-01.ac-cn-hongkong-c.wakuv2.test:~ % sudo tcpdump -n dst port 8000
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
11:57:35.474937 IP 1.2.3.4.54352 > 172.31.239.121.8000: Flags [S], seq 3373374340, win 64240, options [mss 1460,sackOK,TS val 3397820430 ecr 0,nop,wscale 7], length 0
11:57:36.489841 IP 1.2.3.4.54352 > 172.31.239.121.8000: Flags [S], seq 3373374340, win 64240, options [mss 1460,sackOK,TS val 3397821445 ecr 0,nop,wscale 7], length 0
11:57:38.534999 IP 1.2.3.4.54352 > 172.31.239.121.8000: Flags [S], seq 3373374340, win 64240, options [mss 1460,sackOK,TS val 3397823492 ecr 0,nop,wscale 7], length 0
11:57:42.567622 IP 1.2.3.4.54352 > 172.31.239.121.8000: Flags [S], seq 3373374340, win 64240, options [mss 1460,sackOK,TS val 3397827524 ecr 0,nop,wscale 7], length 0
^C
4 packets captured
4 packets received by filter
0 packets dropped by kernel
```
So `websocat` connections just repeats the initial  `SYN` packet.w

**Comment by jakubgs** [comment]
As far as I can tell the `nim-waku` node is in some kind of invalid state that causes it to fail to handle new WebSocket connections.

**Comment by jakubgs** [comment]
I built an image based on `ubuntu:21.10` instead of `ubuntu:20.04` and I got the same `machineId`:
```
 > d run -it --rm -h other-hostname statusteam/squish-licenseserver:v6.7.2
2022-02-22T12:31:11.788 Info: No configuration supplied. Please provide the machineId "123qwe" to our team at sales@froglogic.com to get a valid configuration.
```
So it's not dependent on OS version.

**Comment by jm-clius** [comment]
Mmm. Interesting, my first intuition is that this may be related to https://github.com/status-im/nim-websock/issues/79

But this is triggered by invalid websocket connection attempts, which afaik has not been the case on any of the fleet nodes so far. With port `8000` now reachable on `node-01.do-ams3.wakuv2.test.statusim.net`, I can create and destroy secure websocket connections without any issue:

Using `openssl`:
```shell
openssl s_client -connect node-01.do-ams3.wakuv2.test.statusim.net:8000
CONNECTED(00000003)
...
```

From another `nim-waku` client:
```
INF 2022-02-22 14:17:32.514+01:00 Attempting dial                            topics="wakunode" tid=3398 file=wakunode2.nim:772 wireAddr=/dns4/node-01.do-ams3.wakuv2.test.statusim.net/tcp/8000/wss peerId=16U*o97bUZ
DBG 2022-02-22 14:17:32.514+01:00 Adding peer to manager                     topics="wakupeers" tid=3398 file=peer_manager.nim:192 peerId=16U*o97bUZ addr=/dns4/node-01.do-ams3.wakuv2.test.statusim.net/tcp/8000/wss proto=/vac/waku/relay/2.0.0
INF 2022-02-22 14:17:32.514+01:00 Dialing peer from manager                  topics="wakupeers" tid=3398 file=peer_manager.nim:51 wireAddr=/dns4/node-01.do-ams3.wakuv2.test.statusim.net/tcp/8000/wss peerId=16U*o97bUZ
DBG 2022-02-22 14:17:32.514+01:00 Dialing peer                               topics="libp2p dialer" tid=3398 file=dialer.nim:52 peerId=16U*o97bUZ
DBG 2022-02-22 14:17:33.256+01:00 created new pubsub peer                    topics="libp2p pubsub" tid=3398 file=pubsub.nim:297 peerId=16U*o97bUZ
DBG 2022-02-22 14:17:33.258+01:00 Incoming WakuRelay connection              topics="wakurelay" tid=3398 file=waku_relay.nim:33
DBG 2022-02-22 14:17:33.258+01:00 starting pubsub read loop                  topics="libp2p pubsubpeer" tid=3398 file=pubsubpeer.nim:116 conn=16U*o97bUZ:6214e26d033e9da0201c8045 peer=16U*o97bUZ closed=false
DBG 2022-02-22 14:17:33.482+01:00 decodeMsg: decoded message                 topics="libp2p identify" tid=3398 file=identify.nim:110 pubkey=secp25...63651) addresses="@[/ip4/134.209.139.210/tcp/30303, /ip4/134.209.139.210/tcp/8000/wss]" protocols="@[\"/ipfs/id/1.0.0\", \"/vac/waku/relay/2.0.0\", \"/ipfs/ping/1.0.0\", \"/vac/waku/swap/2.0.0-beta1\", \"/vac/waku/store/2.0.0-beta3\", \"/vac/waku/lightpush/2.0.0-beta1\", \"/vac/waku/filter/2.0.0-beta1\"]" observable_address=Some(/ip4/90.113.197.14/tcp/55460/wss) proto_version="Some(\"ipfs/0.1.0\")" agent_version="Some(\"nim-libp2p/0.0.1\")"
DBG 2022-02-22 14:17:33.482+01:00 starting pubsub read loop                  topics="libp2p pubsubpeer" tid=3398 file=pubsubpeer.nim:116 conn=16U*o97bUZ:6214e26d033e9da0201c8047 peer=16U*o97bUZ closed=false
DBG 2022-02-22 14:17:33.483+01:00 Dial successful                            topics="libp2p dialer" tid=3398 file=dialer.nim:109 conn=16U*o97bUZ:6214e26d033e9da0201c8043 peerId=16U*o97bUZ
INF 2022-02-22 14:17:33.666+01:00 Successfully connected to peer             topics="wakunode" tid=3398 file=wakunode2.nim:776 wireAddr=/dns4/node-01.do-ams3.wakuv2.test.statusim.net/tcp/8000/wss peerId=16U*o97bUZ
```

**Comment by jm-clius** [comment]
Another idea is that something strange may happen because of the reconnection behaviour of fleet nodes. After a restart each node will attempt to reconnect to peers it previously received via the `--staticnode` option or `post_waku_v2_admin_v1_peers` RPC call. No idea if related yet, but basing this on:
1. `16Uiu2HAmPLe7Mzm8TsYUubgCAW1aJoeFScxrLj8ppHFivPo97bUZ` on `node-01.do-ams3.wakuv2.test` tried to reconnect to previous peers after its most recent restart.
2. When it tried to connect `16Uiu2HAkvWiyFsgRhuJEb9JfjYxEkoHLgnUQmr1N5mKWnYjxYRVm` on `node-01.ac-cn-hongkong-c.wakuv2.test` it failed to negotiate a secure connection at the IPv4 address it had stored for that peer.
3. And instead found peer ID `16Uiu2HAm4v86W3bmT1BiH6oSPzcsSr24iDQpSN5Qa992BCjjwgrD`, which is the peer ID for the `prod` fleet node on `node-01.ac-cn-hongkong-c.wakuv2.prod`

In other words, it appears the `Peer` table on the `test` node contains IP addresses for the `prod` nodes. Was there a change in IP allocation that could explain this?

No reason yet to expect this to be related to the websocket issue, but a starting point to investigate as everything else seem to work now (on `node-01.do-ams3.wakuv2.test`)

**Comment by jm-clius** [comment]
Ok, it seems like we hit this issue again. Something causes new `/wss` connections to `node-01.do-ams3.wakuv2.test.statusim.net:8000` now to hang on connection.

Haven't looked at logs yet, but this does some consistent with the blocking caused by https://github.com/status-im/nim-websock/issues/79 (cc @Menduist)

**Comment by jakubgs** [comment]
Indeed, local tests fail too:
```
admin@node-01.do-ams3.wakuv2.test:~ % websocat -vk1E wss://localhost:8000
[INFO  websocat::lints] Auto-inserting the line mode
[INFO  websocat::stdio_threaded_peer] get_stdio_peer (threaded)
[INFO  websocat::ws_client_peer] get_ws_client_peer
websocat: WebSocketError: I/O failure
websocat: error running
```
But the node keeps running fine:
```
admin@node-01.do-ams3.wakuv2.test:~ % d
CONTAINER ID   NAMES         IMAGE                                CREATED       STATUS
bb29852c47fc   websockify    statusteam/websockify:v0.10.0        4 hours ago   Up 4 hours
1c8552337327   nim-waku-v2   statusteam/nim-waku:deploy-v2-test   4 hours ago   Up 4 hours (healthy)
```

**Comment by jakubgs** [comment]
This is getting silly. The nodes stop responding to websocket connections 18 minutes after restarting the node:
```
admin@node-01.do-ams3.wakuv2.test:~ % c | grep waku
nim-waku-v2             nim-waku-v2                tcp     passing
nim-waku-v2-metrics     nim-waku-v2-metrics        http    passing
nim-waku-v2-websocket   nim-waku-v2-websocket      script  critical
nim-waku-v2-websockify  nim-waku-v2-websockify     script  passing

admin@node-01.do-ams3.wakuv2.test:~ % d
CONTAINER ID   NAMES         IMAGE                                CREATED        STATUS
f4d171c92c6b   nim-waku-v2   statusteam/nim-waku:deploy-v2-test   17 hours ago   Up 18 minutes (healthy)
c90ccd9b7fd6   websockify    statusteam/websockify:v0.10.0        5 days ago     Up 34 hours

admin@node-01.do-ams3.wakuv2.test:~ % d restart nim-waku-v2
nim-waku-v2

admin@node-01.do-ams3.wakuv2.test:~ % c | grep waku
nim-waku-v2             nim-waku-v2                tcp     passing
nim-waku-v2-metrics     nim-waku-v2-metrics        http    passing
nim-waku-v2-websocket   nim-waku-v2-websocket      script  passing
nim-waku-v2-websockify  nim-waku-v2-websockify     script  passing
```
The amount of alerts I've received for Nim-Waku websockets in the past week or so is ridiculous:

![image](https://user-images.githubusercontent.com/2212681/156144327-ec5e2d3b-7e9e-49ed-a9c8-78ef4bd00ed0.png)

Until this is resolved I'm going to disable WebSocket healthchecks.

**Comment by jakubgs** [comment]
There you go. I bumped the alert threshold to 2 weeks for now: https://github.com/status-im/infra-nim-waku/commit/91519c24

**Comment by jm-clius** [comment]
Thanks, Jakub. Highest priority after we get the release out.

**Comment by jakubgs** [comment]
Actually, setting `failures_before_critical` doesn't seem to affect script checks: https://github.com/status-im/infra-role-nim-waku/commit/cb2b857b

**Comment by jm-clius** [comment]
May be fixed by https://github.com/status-im/nim-websock/pull/102

**Comment by jakubgs** [comment]
@dryajov maybe you're not aware but we do have a Kibana instance:
https://kibana.infra.status.im/goto/6a42c600-9a68-11ec-8d4d-fdbb12aa0c35

![image](https://user-images.githubusercontent.com/2212681/156445520-9f213995-d661-4445-9bf6-db5d215c9729.png)

**Comment by dryajov** [comment]
> @dryajov maybe you're not aware but we do have a Kibana instance: https://kibana.infra.status.im/goto/6a42c600-9a68-11ec-8d4d-fdbb12aa0c35
> 
> ![image](https://user-images.githubusercontent.com/2212681/156445520-9f213995-d661-4445-9bf6-db5d215c9729.png)

I don't work with waku and I'm not familiar with it's infrastructure, guidance is appreciated - thanks.

**Comment by jakubgs** [comment]
All infra services can be found in `infra-docs` repo: https://github.com/status-im/infra-docs

**Comment by jm-clius** [comment]
@D4nte not sure if you've done or would like to do any more js-waku specific tests since the upgrade to v0.9? My own tests pass and I've been unable to reproduce the hanging websocket issue. Seems the submodule fix worked. LMK if we can close this issue.

**Comment by jm-clius** [comment]
Closing issue, as this has not reoccurred after the fix provided in #905.

**Comment by D4nte** [comment]
Thx, tracking proper testing in https://github.com/status-im/js-waku/issues/517
